storage/rocksdb/clone/donor.cc

```diff
@@ -0,0 +1,1462 @@
+        mysql_mutex_unlock(&m_donor_mutex);
+        assert(erased_count == 1);
+        continue;
+      }
+      MyOsError(errn, EE_FILENOTFOUND, MYF(0), donor_file_path.c_str());
+      save_error(ER_CANT_OPEN_FILE, donor_file_path, errn);
+      return ER_CANT_OPEN_FILE;
+    }
+
+#ifdef __APPLE__
+    if (use_direct_io && fcntl(fd, F_NOCACHE, 1) == -1) {
+      const auto errn = my_errno();
+      // EE_STAT is not an exact match, but this is enough of a corner case
+      MyOsError(errn, EE_STAT, MYF(0), donor_file_path.c_str());
+      save_error(ER_CANT_OPEN_FILE, donor_file_path, errn);
+      mysql_file_close(fd, MYF(MY_WME));
+      return myrocks::clone::return_error(ER_CANT_OPEN_FILE,
+                                          "Failed to fcntl(F_NOCACHE) a file");
+    }
+#endif
+
+    const auto new_pos =
+        my_seek(fd, metadata.get_sent_offset(), SEEK_SET, MYF(MY_WME));
+    if (new_pos != metadata.get_sent_offset()) {
+      assert(new_pos == MY_FILEPOS_ERROR);
+      mysql_file_close(fd, MYF(MY_WME));
+      save_error(ER_FSEEK_FAIL, donor_file_path, my_errno());
+      return myrocks::clone::return_error(ER_FSEEK_FAIL,
+                                          "Failed to seek a file");
+    }
+
+    if (!metadata.started()) {
+      err = send_metadata(metadata, cbk);
+      if (err != 0) return err;
+    }
+
+    Ha_clone_file clone_file{
+#ifndef __APPLE__
+        use_direct_io,
+#else
+        false,
+#endif
+        Ha_clone_file::FILE_DESC,
+        {fd}};
+    cbk.set_source_name(metadata.get_name().c_str());
+
+    auto buf_size = cbk.get_client_buffer_size();
+    const auto remote_clone = buf_size == 0;
+
+    cbk.clear_flags();
+    if (!use_direct_io) cbk.set_os_buffer_cache();
+    if (!remote_clone) cbk.set_zero_copy();
+
+    if (remote_clone) {
+      // 1M value matches InnoDB clone behavior. We could also consider deriving
+      // the value from i.e. max_allowed_packet value.
+      buf_size = 1024 * 1024;
+    }
+    assert(buf_size > 0);
+
+    while (!metadata.eof()) {
+      err = handle_any_error(thd);
+      if (err != 0) {
+        mysql_file_close(fd, MYF(MY_WME));
+        return err;
+      }
+
+      const auto chunk_size = metadata.chunk_size(buf_size);
+      donor_chunk_metadata chunk{metadata.get_id(), chunk_size};
+      myrocks::Rdb_string_writer buf;
+      chunk.serialize(buf);
+      cbk.set_data_desc(buf.ptr(), buf.get_current_pos());
+
+      DBUG_EXECUTE_IF("myrocks_clone_donor_copy_file_crash", DBUG_SUICIDE(););
+
+      err = cbk.file_cbk(clone_file, chunk_size);
+      if (err != 0) {
+        save_error(err, donor_file_path, my_errno());
+        mysql_file_close(fd, MYF(MY_WME));
+        return myrocks::clone::return_error(err, "Failed to send file data");
+      }
+      metadata.advance(chunk_size);
+    }
+
+    mysql_mutex_lock(&m_donor_mutex);
+    auto in_progress_files_itr = m_in_progress_files.find(metadata.get_id());
```

Comment:
You can probably combine this critical section with the one after closing the files, or with the section selecting the file for transfer, to reduce the number of lock requests. There are already codepaths above which immediately return when encountering an error, so whether or not m_completed_files includes files that failed the close is probably a minor point.

Reply:
I have merged this section with the one after closing the file. I am paranoid about doing any I/O under mutex, but in this case it should be fine. I am concerned about extending it further, let's resolve the checkpoint-under-lock issue first.